### ISSUE TITLE
When running odo version, should report server details if when not logged in.

### ISSUE THREAD

**Issue by GrahamDumpleton** [redhat-developer/odo#1028]
[kind/Enhancement]

## Which functionality do you think we should update/improve?

Reporting by ``odo version``.

It doesn't provide server version details when RBAC doesn't permit access.

```
[user4:~] $ oc whoami
Error from server (Forbidden): users.user.openshift.io "~" is forbidden: User "system:anonymous" cannot get users.user.openshift.io at the cluster scope: no RBAC policy
matched
```

At the same time, ``oc`` and ``kubectl`` have no issues.

```
[user4:~] $ oc version
oc v3.11.44
kubernetes v1.11.0+d4cacc0
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://172.30.0.1
openshift v3.11.16
kubernetes v1.11.0+d4cacc0
[user4:~] $ echo $?
0
[user4:~] $ kubectl version
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-
27T20:17:28Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11+", GitVersion:"v1.11.0+d4cacc0", GitCommit:"d4cacc0", GitTreeState:"clean", BuildDate:"2018-09-26T12:30:09Z", GoVersion
:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
[user4:~] $ echo $?
0
[user4:~] $ odo version
odo v0.0.14 (618c6d3)
Unable to connect to OpenShift cluster, is it down?
[user4:~] $ echo $?
1
```

## Why is this needed? 

Would be good to see it be consistent and that it can work like ``oc`` and ``kubectl``.

It could be a notable issue in that ``odo version`` in this case returns an exit error code for program. That could screw up scripts that want to invoke ``odo version``.

**Comment by syamgk** [comment]
This was fixed in https://github.com/redhat-developer/odo/commit/16b54740bd62dfeab53c98400fa0fcd149d8d32f

I can see you are on Odo version  v0.0.14
this fix came out on v0.0.15

**Comment by GrahamDumpleton** [comment]
Oh crap. I was supposed to be using the latest. I must have stuffed things up when I was changing the build for this image which was bundling ``odo``. Thanks.

**Comment by GrahamDumpleton** [comment]
Reopening this issue. The variation I am testing isn't fixed in 0.0.15.

```
[user5:~] $ oc version
oc v3.11.44
kubernetes v1.11.0+d4cacc0
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://172.30.0.1
openshift v3.11.16
kubernetes v1.11.0+d4cacc0
[user5:~] $ odo version
odo v0.0.15 (2f7ed497)
Unable to connect to OpenShift cluster, is it down?
[user5:~] $ echo $?
1
```

**Comment by syamgk** [comment]
It seems its working for me,
it will be helpful if you can you share the log of `odo version -v 9`?

**Comment by GrahamDumpleton** [comment]
Seems to die because doesn't have explicit port specified, which it should be able to infer from fact ``https`` is used.

```
[user3:~] $ oc version
oc v3.11.44
kubernetes v1.11.0+d4cacc0
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://172.30.0.1
openshift v3.11.16
kubernetes v1.11.0+d4cacc0

[user3:~] $ kubectl version
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:17:28Z", GoVersion:"go1.10.2
", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11+", GitVersion:"v1.11.0+d4cacc0", GitCommit:"d4cacc0", GitTreeState:"clean", BuildDate:"2018-09-26T12:30:09Z", GoVersion:"go1.10.2", Compiler:"gc", Platfo
rm:"linux/amd64"}

[user3:~] $ odo version -v 9
I1122 07:44:27.236905     250 version.go:40] KUBECTL_VERSION=1.11+
odo v0.0.15 (2f7ed497)
I1122 07:44:27.238226     250 loader.go:359] Config loaded from file /opt/app-root/src/.kube/config
I1122 07:44:27.239057     250 occlient.go:365] Trying to connect to server 172.30.0.1
I1122 07:44:27.239104     250 occlient.go:368] unable to connect to server: dial tcp: address 172.30.0.1: missing port in address
I1122 07:44:27.239124     250 client.go:29] Error:
Unable to connect to OpenShift cluster, is it down?
Unable to connect to OpenShift cluster, is it down?
```

**Comment by GrahamDumpleton** [comment]
Note that the twist in this is that is after having set up a custom context. Doing that as doing this inside of a container in OpenShift.

```
# Setup 'oc' client configuration for the location of the OpenShift cluster.

CA_FILE="/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"

if [ x"$KUBERNETES_PORT_443_TCP_ADDR" != x"" ]; then
    if [ -f $CA_FILE ]; then
        OC_CA_ARGS="--certificate-authority $CA_FILE"
    else
        OC_CA_ARGS="--insecure-skip-tls-verify"
    fi

    oc config set-cluster local $OC_CA_ARGS \
            --server "https://$KUBERNETES_PORT_443_TCP_ADDR" 

    oc config set-context local --cluster local
    oc config use-context local
fi
```

If I add the port, then it doesn't fail in the same way, but still doesn't give server version because it doesn't then like is a self signed certificate.

```
[user3:~] $ oc config set-cluster local $OC_CA_ARGS --server "https://$KUBERNETES_PORT_443_TCP_ADDR:443"
Cluster "local" set.
[user3:~] $ odo version
odo v0.0.15 (2f7ed497)
Get https://172.30.0.1:443/version: x509: certificate signed by unknown authority
```

It should default the port to 443 if ``https`` is used. And for getting server version, it would be better to allow self sign certificate.

**Comment by GrahamDumpleton** [comment]
Or should say, should honour the certificate authority from the context in this case.

Even then, for ``oc version``, it perhaps shouldn't grumble about self signed certificate either.

**Comment by GrahamDumpleton** [comment]
Ignore comments about certificate authority, it is just the port, and potentially self signed certificate, that is the issue. In the latest test didn't set up the arguments properly to have certificate authority set on the cluster context.

**Comment by GrahamDumpleton** [comment]
I am going to add the port when setting the context in my stuff. Will at least get around the issue. Still feel that ``odo`` should default the port to 443 if is ``https`` though and not get confused.

**Comment by kadel** [comment]
> I am going to add the port when setting the context in my stuff. Will at least get around the issue. Still feel that `odo` should default the port to 443 if is `https` though and not get confused.

You are right that is bug, odo should do that.

We didn't notice this before because if you use `oc login` or `odo login` than the port in kubeconfig gets automatically populated even if you don't specify it in your command (`oc login https://mycluster.com` will add `server: https://mycluster.com:443` to kubeconfig)

**Comment by kadel** [comment]
part with default ports will be fixed in #1030 

but we still need to address the issue with the exit code


`odo version` should not exit with error code when it can't connect to cluster, and the error message should be probably hidden

**Comment by kadel** [comment]
reopening this as we also need https://github.com/redhat-developer/odo/pull/1030 to fix all issues described here
